web/internal/metadata.bzl

```diff
@@ -0,0 +1,153 @@
+  if browser_label:
+    content += ',\n  "browserLabel": "%s"' % browser_label
+  if test_label:
+    content += ',\n  "testLabel": "%s"' % test_label
+  if web_test_files:
+    content += ',\n  "webTestFiles": ['
+    first = True
+    for f in web_test_files:
+      if first:
+        first = False
+      else:
+        content += ','
+      content += '\n' + _web_test_files_to_string(ctx, f, '    ')
+    content += '\n  ]'
+  content += '\n}\n'
+
+  ctx.file_action(output=output, content=content, executable=False)
+
+
+def _get_files(ctx, attr_names):
+  """Finds all of the web_test_metadata files in given attributes.
+
+  It ensures that all web_test_metadata files position in the returned list is
+  based on the last attribute in attr_names that they appear in.
+
+  Args:
+    ctx: the context object for this rule.
+    attr_names: the names of attributes to get web_test_metadata files from.
+  Returns:
+    a list of all of the web_test_metadata files in the specificied attributes.
+  """
+  metadata_files = []
+
+  for attr_name in attr_names:
+    if hasattr(ctx.attr, attr_name):
+      attr = getattr(ctx.attr, attr_name)
+      if lists.is_list_like(attr):
+        for item in attr:
+          _get_files_helper(metadata_files, item)
+      else:
+        _get_files_helper(metadata_files, attr)
+  return metadata_files
+
+
+def _get_files_helper(metadata_files, r):
+  if hasattr(r, 'transitive_web_test_metadata'):
+    lists.ensure_all_at_end_of_list(metadata_files,
+                                    r.transitive_web_test_metadata)
+  if hasattr(r, 'web_test_metadata'):
+    lists.ensure_all_at_end_of_list(metadata_files, r.web_test_metadata)
+
+
+def _web_test_files(named_files={}, archive_file=None):
+  return struct(archive_file=archive_file, named_files=named_files)
+
+
+def _web_test_files_to_string(ctx, files_struct, indent=''):
+  """Converts a web_test_files struct to a JSON string."""
+  result = indent + '{\n'
+  if files_struct.archive_file:
+    result += indent + '  "archiveFile": "' + files.runfiles_path(
+        ctx, files_struct.archive_file) + '",\n'
+  result += indent + '  "namedFiles": {'
+  first = True
+  for n, f in files_struct.named_files.items():
+    if first:
+      first = False
+    else:
+      result += ','
+    result += '\n' + indent + '    "' + n + '": "'
+    if type(f) == type(''):
+      result += f
+    else:
+      result += files.runfiles_path(ctx, f)
+    result += '"'
+  result += '\n' + indent + '  }\n' + indent + '}'
+  return result
+
+
+def _aspect_impl(target, ctx):
+  transitive_files = _get_files(ctx.rule, ['data', 'deps', 'srcs'])
+
+  if hasattr(target, 'web_test_metadata'):
+    for m in target.web_test_metadata:
+      lists.ensure_at_end_of_list(transitive_files, m)
+
+  return struct(transitive_web_test_metadata=transitive_files)
+
+
+_aspect = aspect(
+    implementation=_aspect_impl,
+    attr_aspects=['deps', 'data', 'srcs', 'browser', 'config'],)
```

Comment:
Aspects are extremely expensive. More so when it's traversing all common properties. It causes every single rule in the transitive closure to be re-evaluated. Traditionally the cost of constructing the build graph for a repository is `O(#rules)`. With aspects it's `O(#rules * #aspects)`. With a repository as large as the one we have internally, it becomes clear why the introduction of new aspects is something that merits careful consideration.

Are you sure an aspect is what you want? What's the use case?

It seems like the aspect exists in order to aggregate the `web_test_metadata` provider exported by your rules. But couldn't this be done by simply having the skylark rules themselves aggregate this data from child rules, and then re-exporting it to parent rules via providers? See the `collect_js` function in `defs.bzl` in cl/136947978 for an example of this.

The only use case I know of for aspects is that they let you bolt on functionality to rules other people have written. For example, cl/136947978 has an aspect defined in legacy_js.bzl. It basically causes `closure_js_library` to crawl all transitive `deps` looking for `js_library` rules. Then it runs JsChecker to validate their sources, and finally adds a `closure_js_library` provider to those rules so they become polymorphically compatible with `closure_js_library` and therefore satisfy the `"deps": attr.label_list(providers=["closure_js_library"])` precondition. TypeScript uses aspects to run Clutz and tsickle on js_library sources so they can be linked into TypeScript applications. Bazel team is also planning on using aspects to build language-specific bindings for proto_library().